### PR TITLE
Adds the possibility to input a regular expression to the scrapper call

### PR DESCRIPTION
**Pull request by dpshelio**
### Description
If the URL that you want to scrap has some characters that change but not follow to any particular datetime pattern, then now you can use regex on the url:

Inspired by the imagination of gong synoptic maps archive (#4055)

```python
from sunpy.util.scraper import Scraper 
from sunpy.time import TimeRange 
pattern = r'https://gong2.nso.edu/oQR/zqs/%Y%m/mrzqs%y%m%d/mrzqs%y%m%dt%H%Mc(\d){4}_(\d){3}\.fits.gz'
scraper = Scraper(pattern, regex=True)
timerange = TimeRange('2020-01-05','2020-01-06T16:00:00')
scraper.filelist(timerange)
```

### CODE REVIEW

**Comment by abhijeetmanhas** [comment]
This will also help to solve #3337

File: changelog/4107.feature.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Adds to ~sunpy.net.scraper.Scraper~ the ability to include regular expressions in the URL passed.
```
Comment (Cadair) [inline changelog/4107.feature.rst:1]:
```suggestion
Adds to `~sunpy.net.scraper.Scraper` the ability to include regular expressions in the URL passed.
```

**Comment by Cadair** [comment]
This looks good, and future improvements shouldn't stop this as it has short term uses.

For things like #3337 we are also going to need the ability to capture these regex. Also future tighter integration of `GenericClient` with scraper makes me wonder if something like https://github.com/r1chardj0n3s/parse would be better than the wonderful world of regex capture groups?

**Comment by dpshelio** [comment]
> For things like #3337 we are also going to need the ability to capture these regex. Also future tighter integration of `GenericClient` with scraper makes me wonder if something like https://github.com/r1chardj0n3s/parse would be better than the wonderful world of regex capture groups?

Yeah! redesign scraper with `parser` will be an improvement!

**Comment by hayesla** [comment]
Yay - it has seriously bugged me not having this in the past!

I'm also all in favor of booting regex

**Comment by dpshelio** [comment]
Let's see how the `urlparser` handles regex in between....

File: sunpy/util/scraper.py [hunk 2]
```diff
@@ -61,8 +68,13 @@ class Scraper:
     The ``now`` attribute does not return an existent file, but just how the
     pattern looks with the actual time.
     """
-    def __init__(self, pattern, **kwargs):
-        self.pattern = pattern.format(**kwargs)
+    def __init__(self, pattern, regex=False, **kwargs):
+        if regex:
+            self.pattern = pattern
+            if kwargs:
+                warnings.warn('regexp being used, the extra arguments passed are being ignored')
```
Comment (hayesla) [inline sunpy/util/scraper.py:75]: should this also use a `SunpyUserWarning`?

 i.e. 
`warnings.warn('regexp being used, the extra arguments passed are being ignored', SunpyUserWarning )`

Comment (dpshelio) [reply sunpy/util/scraper.py]: Thanks!

**Comment by nabobalis** [comment]
```
ERROR ../../.tox/py38-online/lib/python3.8/site-packages/sunpy/net/helio/tests/test_helio.py::test_get_table_names - requests.exceptions.ConnectionError: HTTPConnectionPool(host...
ERROR ../../.tox/py38-online/lib/python3.8/site-packages/sunpy/net/helio/tests/test_helio.py::test_select_table - requests.exceptions.ConnectionError: HTTPConnectionPool(host='w...
ERROR ../../.tox/py38-online/lib/python3.8/site-packages/sunpy/net/helio/tests/test_helio.py::test_time_query - requests.exceptions.ConnectionError: HTTPConnectionPool(host='www...
```
Merging.

**Pull request merged**